Proposal containing the results of the SilkX experiment (#1727)
dotnet/Silk.NET: documentation/proposals/Proposal - Generation of Library Sources and PInvoke Mechanisms.md

Tanner has already stated that they're happy to work with us to get Silk.NET using ClangSharp, even saying they're happy to add a CI test stage into the ClangSharp repo to ensure no incoming changes break Silk.NET's generation process.

# Implicit Cast-Based Overloading

The pivotal proposal around which the SilkX experiment revolves is solving the excessive overload problem by having no overloads at all. Namely, Silk.NET.Core will have a set of "pointer-like" types that can be implicitly casted to by a multitude of different types. This gives the user the same flexibility as having all different permutations of overloaded parameters generated without having to do so, and thereby alleviates the requirement of having overloads at all.

> `Ptr` and `Mut` `ref struct`s **shall** be defined wrapping a `ref` and `ref readonly` respectively. These types **shall** have generic and non-generic variants. This shall be repeated up to 3 pointer dimensions, taking into account C's mutability rules in each instantiation. This means that there will be 14 top-level generic structs and 14 top-level non-generic structs. These **should** be auto-generated to ensure they remain maintainable. IL weaving **may** be required for the higher dimensions.

Reviewer: backwards

Author: ```suggestion
`Ptr` and `Mut` `ref struct`s **shall** be defined wrapping a `ref readonly` and `ref` respectively. These types **shall** have generic and non-generic variants. This shall be repeated up to 3 pointer dimensions, taking into account C's mutability rules in each instantiation. This means that there will be 14 top-level generic structs and 14 top-level non-generic structs. These **should** be auto-generated to ensure they remain maintainable. IL weaving **may** be required for the higher dimensions.
```

> Unless deemed inappropriate, inapplicable, and/or infeasible by the Silk.NET team (the team reserved the right to do so without Working Group approval), the types **should** have the following characteristics:

Author: In the case of a generic pointer type, implicitly convertible to its non-generic counterpart?

Reviewer 2: This should be easy enough as we would just be passing the InteriorRef as the Ref for the non-generic consturctor.

> - An instance **must** be constructable from a `ref T` or `ref readonly T` (depending on the outer dimension mutability) where `T` is either the generic type (or `byte` for non-generic variants), a "pointer-like" type of the inner dimension of type being constructed.

Reviewer 2: Regarding dimension mutability, more explicit definition maybe good. As of right now inner-most is first and outermost is last. This may also not be the most intuitive as I read it in the reverse. So MutPtr<byte> would be byte** const rather than byte * const *.

> - An indexer **must** be present accepting a `nuint` for the index, returning a `ref T` or `ref readonly T` (depending on the outer dimension mutability) where `T` is either the generic type (or `byte` for non-generic variants), a "pointer-like" type of the inner dimension of type being indexed.

Reviewer: clarify ability to return other Ptr-like types

Author: I thought this was implied by `inner dimension`, but alright will reword

Reviewer: yeah i just wanted that clarified in the wording, i think the parens around the `byte`/non-generic version hides it a bit at least to my reading

Author: ```suggestion
- An indexer **must** be present accepting a `nuint` for the index, returning a `ref T` or `ref readonly T` (depending on the outer dimension mutability) where `T` is either the generic type (or `byte` for non-generic variants), a "pointer-like" type of the inner dimension (i.e. a `PtrPtr<T>`'s indexer will return a `Ptr<T>`) of type being indexed.
```

> - An indexer **must** be present accepting a `nuint` for the index, returning a `ref T` or `ref readonly T` (depending on the outer dimension mutability) where `T` is either the generic type (or `byte` for non-generic variants), a "pointer-like" type of the inner dimension of type being indexed.

Reviewer: permit multiple-argument indexers on higher-arity Ptr types

- For single-dimension pointers, an `AsSpan` method **must** be present taking an `int` length argument (due to historical reasons in the .NET BCL) returning a `Span<T>` or `ReadOnlySpan<T>` (depending on the outer dimension mutability) where `T` the pointee type. For non-generic pointer-like types, a generic argument **may** be used to specify the type of the span.
- An array of the same jagged dimension **must** be implicitly convertible to the type.
- For single-dimension pointers, a multi-dimensional array (up to 3 dimensions) **must** be implicitly convertible to the type.
- A raw pointer of the same dimensions **must** be implicitly convertible to the type.
- `NullPtr` **must** be implicitly convertible to the type, and the returned pointer must represent a null reference.
- An explicit operator **must** be present to unsafely convert the reference represented by the pointer to a raw pointer without the user using `fixed`/`GetPinnableReference`.
- An explicit operator **must** be present to unsafely convert the reference represented by the pointer to a void pointer without the user using `fixed`/`GetPinnableReference`.

> - For generic pointer types, an implicit operator **must** be present to convert the pointer type to a `string` (or an array of strings of the inner dimension's jaggedness i.e. `PtrPtrPtr<byte>` becomes `string[][]`). Much like `ReadOnlySpan<char>`, `ToString` **must** call this implicitly if the type is a single-dimension pointer to a string pointee type. The implicit cast **may** throw an exception if the pointee type is not a string pointee type. This is because we can't constrain the type used on implicit operators on generic types.

Reviewer: still feel uneasy with implicit string conversions for allocation/perf reasons but that would need benchmarking and further analysis and consideration

Author: If you don't want to marshal a string just don't pass a string 🤷

Reviewer 2: Maxine brought this up in the discord, this may be better as an explicit operator as assuming a Ptr<T> equates to a string is dangerous.

Reviewer 2: It is also probably good to limit this functionality to only String types (byte, sbyte, ushort, short, uint, int). Using extensions would work for a method, but we can't directly override an operator. It may be best to use an explicit function like ReadToString(), then the operator can call the function and we override that for specific types otherwise throw an exception.

Author: Agreed, I think we could define `ReadToString` in `SilkMarshal` as an extension method for `Ptr`, `Mut`, `Ptr<T>`, and `Mut<T>` where T is one of: byte, sbyte, ushort, short, char, uint, int

    }

    // Non-Static Interface
    private Func<string, nint> _getProcAddress;
    byte* IMyStringLibrary.ToLower(byte* str)
    {
        var ptr = _getProcAddress("ToLower");
        if (ptr is 0) throw new("some symbol loading exception...");
        return ((delegate* unmanaged<byte*, byte*>)ptr)(str);
    }
    
    void IMyStringLibrary.FreeResult(byte* str)
    {
        var ptr = _getProcAddress("FreeResult");
        if (ptr is 0) throw new("some symbol loading exception...");
        return ((delegate* unmanaged<byte*, void>)ptr)(str);
    }

    public static IMyStringLibrary Create() => new StaticWrapper<DllImport>();
    public static IMyStringLibrary Create(Func<string, nint> getProcAddress) => new MyStringLibrary { _getProcAddress = getProcAddress };

    // Static Interface
    public static byte* ToLower(byte* str) => DllImport.ToLower(str);
    public static Ptr<byte> ToLower(Ptr<byte> str) => DllImport.ToLower(str);
    public static void FreeResult(byte* str) => DllImport.FreeResult(str);
    public static void FreeResult(Ptr<byte> str) => DllImport.FreeResult(str);
}
```

> **INFORMATIVE:** `getProcAddress` delegates replace "native contexts".

Author: Maxine has expressed interest in Discord in not doing this, and maintaining an INativeContext interface in some form rather than forcing delegates to be used.

Unless `ThreadLocal` is itself the *static default* configured by the generator user, the thread-local value within `ThreadLocal` should be instantiated with a value factory calling the `Create` method. If `ThreadLocal` is the *static default*, then no value factory is provided and the user must set the value using `MakeCurrent`.

The binding class **shall** expose a `static` `Create` method with a `Func<string, nint>` parameter returning an instance of the top-level interface. This **shall** use an instance of the binding class itself.

**INFORMATIVE:** It is undecided whether we want SilkTouch to output the `Create` methods itself or whether we want it in a non-generated partial.

The binding class **shall** itself implement the top-level interface, using the `getProcAddress` delegate stored from the `Create` method to implement the native calls using a direct function pointer call.

The binding class **shall** also contain shorthand functions for calling the static functions contained within the static subinterface using the *static default*.

> **INFORMATIVE:** It is undecided whether we want to implement the static subinterface on the binding class itself as well.

Reviewer: I vote yes to this

Author: Yeah fair, just wasn't sure whether we wanted what is essentially a proxy to implement the interface itself, but yeah I don't see why not.

Author: ```suggestion
The binding class **shall** implement the static subinterface (i.e. to proxy calls to the *static default* for ease of use).
```
